### ISSUE TITLE
LoopWithinContext => How to get the root object

### ISSUE THREAD

**Issue by Barre14** [WorkMaze/JUST.net#116]
Hi, 
I've got a array with elements fo different types. Some are of type MainType and some are of type SubType. Each subtype has a property with the id of the maintype.

What i'm trying to do is loop over the maintypes en then create a property SubTypes that holds the subtypes for this specific maintype.

But when using the loopWithinContext, the $ does not point to the root object anymore. Correct?

this is the json input:

```
{
   "ontologyElements":[
      {
         "id":"8b8b9d6e-4574-466b-b2c3-0062ad0642fe",
         "name":"Ontology1",
         "description":"test1",
         "businessKey":"test1",
         "entityType":"**Ontology**",
         "partitionKey":"8b8b9d6e-4574-466b-b2c3-0062ad0642fe"
      },
      {
         "id":"3ac89bbd-0de2-4692-a077-1d5d41efab69",
         "name":"MainType1",
         "order":1,
         "businessKey":"test2",
         "entityType":"**MainType**",
         "partitionKey":"8b8b9d6e-4574-466b-b2c3-0062ad0642fe",
         "ontologyId":"8b8b9d6e-4574-466b-b2c3-0062ad0642fe"
      },
      {
         "id":"97aa4eb2-0515-43d6-ba59-ffd931956b1a",
         "name":"SubType1",
         "order":1,
         "businessKey":"test2",
         "entityType":"**SubType**",
         "partitionKey":"8b8b9d6e-4574-466b-b2c3-0062ad0642fe",
         "ontologyId":"8b8b9d6e-4574-466b-b2c3-0062ad0642fe",
         "mainTypeId:"3ac89bbd-0de2-4692-a077-1d5d41efab69"
      }
   ]
}
```

this is my schema:

```
{
    "id": "#valueof($.ontologyElements[?(@.entityType == 'Ontology')].id)",
    "description": "#valueof($.ontologyElements[?(@.entityType == 'Ontology')].description)",
    "maintypes": {
        "#loop($.ontologyElements[?(@.entityType == 'MainType')])" : {
            "id": "#currentvalueatpath($.id)",
            "name": "#currentvalueatpath($.name)",
            "order": "#currentvalueatpath($.order)",
            "subTypes" : {
                "#loopwithincontext(????????)": {
                    "name": "#currentvalueatpath($.name)"
                    "order": "#currentvalueatpath($.order)"
                  }
            }
        }
    }
}

```
I can not figure out what to put in place of the ?????? so I can loop over all subtypes, select the ones with a maintypeid equal to the current MainType of the first loop.

Can anyone help me?

Bart

**Comment by Courela** [comment]
I don't think it could be done with #loopwithincontext, the jsonpath expression inside it must be a static one, and in this case a dynamic one is needed.

I already made a pull request (#112) about loops, it will remove #loopwithincontext and allow any number of loops, however it will not allow dynamic expressions inside #loop, I'll have to work on that.

**Comment by Courela** [comment]
This works with version 4,0:

```JSON
{
  "id": "#valueof($.ontologyElements[?(@.entityType == 'Ontology')].id)",
  "description": "#valueof($.ontologyElements[?(@.entityType == 'Ontology')].description)",
  "maintypes": {
    "#loop($.ontologyElements[?(@.entityType == 'MainType')])": {
      "id": "#currentvalueatpath($.id)",
      "name": "#currentvalueatpath($.name)",
      "order": "#currentvalueatpath($.order)",
      "subTypes": {
        "name": "#valueof(#xconcat($.ontologyElements[?/(@.entityType == 'MainType' && @.ontologyId == ', #currentvalueatpath($.ontologyId),'/)].name))",
        "order": "#valueof(#xconcat($.ontologyElements[?/(@.entityType == 'MainType' && @.ontologyId == ', #currentvalueatpath($.ontologyId),'/)].order))"
      }
    }
  }
}

```

**Comment by Barre14** [comment]
Wow! Nice. I'm gonna try this!

**Comment by Barre14** [comment]
Hey, I got to test it today and this works great !
Thank you very much !!

**Comment by Barre14** [comment]
Hi,

I was too quick to say that this works :).

It almost works, but consider this schema:

{
    "id": "#valueof($.ontologyElements[?(@.entityType == 'Ontology')].id)",
    "name": "#valueof($.ontologyElements[?(@.entityType == 'Ontology')].name)",
    "description": "#valueof($.ontologyElements[?(@.entityType == 'Ontology')].description)",
    "mainTypes": {
        "#loop($.ontologyElements[?(@.entityType == 'MainType')])": {
            "id": "#currentvalueatpath($.id)",
            "name": "#currentvalueatpath($.name)",
            "order": "#currentvalueatpath($.order)",
            "subTypes": {
                "#loop(#xconcat($.ontologyElements[?/(@.entityType == 'SubType' && @.mainTypeId == ', #currentvalueatpath($.ontologyId),'/)]))": {
                    "id": "#currentvalueatpath($.id)",
                    "name": "#currentvalueatpath($.name)",
                    "order": "#currentvalueatpath($.order)"
                }
            }
        }
    }
}

The difference with your "working" shema is that the loop to find the subtypes does a match on entityType and then on mainTypeId and not on ontologyId. When I changed this, it crashed.

I think this is cause the property mainTypeId is only found on the ones with entitytype "SubType".
Therefore it will not find the property on the other objects and throw a Object reference not set... exception.

Am I correct, that this might be happening with the fix you wrote? Is there a way for you to write this, so that a property that is not found is considered not a match for the loop?

Please let me know, because we were almost there!!
Thank you very much!

Barre14
